editor/src/core/model/element-metadata-utils.ts

```diff
@@ -1202,176 +1145,254 @@ export const MetadataUtils = {
+      elementToInsert.templatePath,
+      withNewElement,
+      indexPosition,
+    )
+    return jsxMetadata(metadata.components, updatedElements)
   },
   duplicateElementMetadataAtPath(
     oldPath: TemplatePath,
     newPath: TemplatePath,
     newElement: Either<string, JSXElementChild>,
-    scenes: Array<ComponentMetadata>,
-  ): Array<ComponentMetadata> {
+    metadata: JSXMetadata,
+  ): JSXMetadata {
+    let workingElements = { ...metadata.elements }
+
+    function duplicateElementMetadata(
+      element: ElementInstanceMetadata,
+      pathToReplace: InstancePath,
+      pathToReplaceWith: InstancePath,
+      newElementInner: Either<string, JSXElementChild>,
+    ): InstancePath {
+      const children = MetadataUtils.getImmediateChildren(metadata, element.templatePath)
+      const duplicatedChildren = children.map((child) => {
+        const childsElement = child.element
+        let duplicatedElement: Either<string, JSXElementChild>
+        if (isLeft(childsElement) || isLeft(newElementInner)) {
+          duplicatedElement = childsElement
+        } else {
+          const childElementUID = getUtopiaID(childsElement.value)
+          duplicatedElement = isJSXElement(newElementInner.value)
+            ? right(newElementInner.value.children.find((c) => getUtopiaID(c) === childElementUID)!)
+            : childsElement
+        }
+        return duplicateElementMetadata(child, pathToReplace, pathToReplaceWith, duplicatedElement)
+      })
+
+      const newTemplatePath = TP.replaceIfAncestor(
+        element.templatePath,
+        pathToReplace,
+        pathToReplaceWith,
+      )
+
+      const newElementMetadata: ElementInstanceMetadata = {
+        ...element,
+        templatePath: newTemplatePath,
+        element: newElementInner,
+        children: duplicatedChildren,
+      }
+
+      workingElements[TP.toString(newTemplatePath)] = newElementMetadata
+      return newTemplatePath
+    }
+
     // Everything about this feels wrong
-    const originalMetadata = getSceneMetadataOrElementInstanceMetadata(oldPath, scenes)
+    const originalMetadata = getSceneMetadataOrElementInstanceMetadata(oldPath, metadata)
     if (originalMetadata == null) {
-      return scenes
+      return metadata
     } else if (isLeft(originalMetadata) && TP.isScenePath(newPath)) {
       const componentMetadata = originalMetadata.value
-      return [
-        ...scenes,
+      // FIXME I think this is wrong and we need to duplicate the children
+      const updatedScenes: ComponentMetadata[] = [
+        ...metadata.components,
         {
           ...componentMetadata,
           scenePath: newPath,
         },
       ]
+      return jsxMetadata(updatedScenes, metadata.elements)
     } else if (
       isRight(originalMetadata) &&
       TP.isInstancePath(oldPath) &&
       TP.isInstancePath(newPath)
     ) {
-      const duplicatedMetadata = this.duplicateElementMetadata(
+      const duplicatedElementPath = duplicateElementMetadata(
```

Comment:
this const doesn't seem to be used

Reply:
Oops good catch. I should strictly have been using this rather than `newPath` in the function call immediately below (the two end up being equal anyway, but I have fixed the code below to correctly use `duplicatedElementPath` in case we change anything internal here later).